### PR TITLE
Move Wasm logic to its own crate

### LINKED ISSUE #2642: Feature-gate Testing Extension
Add `cfg(oak-testing)` and use it instead of [`cfg(test)`](https://github.com/project-oak/oak/blob/910bf1a7b8c036099213bc6c527c63986d6a8b97/oak_functions/loader/src/lib.rs#L31) for the testing extension to allow using the testing extension in integration tests without exposing it by default.

### PR DESCRIPTION
**Pull request by conradgrobler**
This change moves the Wasm handling functionality of the Oak Functions runtime into a separate crate that is `no_std` compatible so that it could be reused by other runtimes (e.g. UEFI).

Fixes #2642 by moving the testing extension to a separate crate.

### CODE REVIEW

File: oak_functions/logger/Cargo.toml [hunk 1]
```diff
@@ -6,4 +6,5 @@ edition = "2021"
 license = "Apache-2.0"
 
 [dependencies]
-log = "*"
+# Logging via the log crate is disabled and only available using our own custom logger.
```
Comment (mariaschett) [inline oak_functions/logger/Cargo.toml:9]: Maybe add link to our own custom logger (trait) and add why we disable logging via log crate?

File: oak_functions/wasm/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+where
+    L: OakLogger,
+{
+    pub fn new(
+        module: &wasmi::Module,
+        request_bytes: Vec<u8>,
+        logger: L,
+        extensions: HashMap<ExtensionHandle, Box<dyn OakApiNativeExtension>>,
+    ) -> anyhow::Result<Self> {
+        let mut abi = Self {
```
Comment (mariaschett) [inline oak_functions/wasm/src/lib.rs:77]: nit: I am not sure why we call this `abi` and not `wasm_state`?

**Review by mariaschett — APPROVED** [review]
Thank you, Conrad! Very nice!

I assume you  just moved and didn't touch anything internal in the WasmState logic I should have a closer look at. Otherwise please let me know!

Also: Thank you for extension-ifying the testing extension.

File: oak_functions/loader/src/server.rs [hunk 3]
```diff
@@ -13,39 +13,14 @@
...
-// Type aliases for positions and offsets in Wasm linear memory. Any future 64-bit version
-// of Wasm would use different types.
-pub type AbiPointer = u32;
-pub type AbiPointerOffset = u32;
-// Type alias for the ExtensionHandle type, which has to be cast into a ExtensionHandle.
-pub type AbiExtensionHandle = i32;
-/// Wasm type identifier for position/offset values in linear memory. Any future 64-bit version of
-/// Wasm would use a different value.
-pub const ABI_USIZE: ValueType = ValueType::I32;
+pub type WasmHandler = oak_functions_wasm::WasmHandler<Logger>;
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_functions/loader/src/server.rs:24]: WasmHandler was made generic to support moving it out of the runtime. This creates a non-generic type alias to be used elsewhere (e.g. in the tests in the example modules) to minimise the impact ot this change.

I have create #2789 to track refactoring and renaming this.

File: oak_functions/logger/Cargo.toml (hunk 1 above)
Comment (conradgrobler) [reply oak_functions/logger/Cargo.toml]: Expanded the comment and added a link to the trait and a TODO.

File: oak_functions/testing/src/lib.rs [hunk 4]
```diff
@@ -47,30 +50,65 @@ impl OakApiNativeExtension for TestingExtension<Logger> {
...
+where
+    L: OakLogger,
+{
+    pub fn new(logger: L) -> Self {
         Self { logger }
     }
 }
+
+#[derive(Clone)]
+pub struct TestingLogger {}
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_functions/testing/src/lib.rs:91]: Added a comment describing the intended use.

**Review by conradgrobler — COMMENTED** [review]
Thanks for the review.

> I assume you just moved and didn't touch anything internal in the WasmState logic I should have a closer look at. Otherwise please let me know!

I made the WasmHandler and WasmState generic to support moving it to a separate crate and split the tests into a separate file. I also moved two functions (`new` and `invoke`) to a different location in the file to avoid having two separate `impl<L> WasmState<L>` blocks.

**Comment by mariaschett** [comment]
> Thanks for the review.
> 
> > I assume you just moved and didn't touch anything internal in the WasmState logic I should have a closer look at. Otherwise please let me know!
> 
> I made the WasmHandler and WasmState generic to support moving it to a separate crate and split the tests into a separate file. I also moved two functions (`new` and `invoke`) to a different location in the file to avoid having two separate `impl<L> WasmState<L>` blocks.

Great! Thank you for the clarification!

**Pull request merged**